### ISSUE TITLE
Evaluate Unsloth as an Atlas service — local fine-tuning, three source flavours, gated on GPU admission control

### ISSUE THREAD

**Issue by thekaveh** [thekaveh/genai-vanilla-stack#937]
Feasibility study for adding [Unsloth](https://github.com/unslothai/unsloth) (71.5k★, Apache-2.0 core / AGPL-3.0 Studio UI) as an Atlas service. Everything below was verified against upstream rather than recalled — see the evidence log at the end.

**No implementation, no decision taken.** This is a study plus a proposed shape, with five decisions that need an owner.

## 1. Recommendation

Model it on **ComfyUI** — the service in Atlas that already has this exact problem — with **three source flavours**, and make **Unsloth Studio** the training surface rather than a footnote.

| Source | Who owns the process | Analogue |
|---|---|---|
| `localhost` | You already run Unsloth; Atlas wires the endpoint | `comfyui: localhost`, `ollama-localhost` |
| `managed-localhost` | Atlas installs, pins, supervises, removes | `comfyui: managed-localhost-mps`, `vllm-metal` |
| `container-gpu` | Linux + NVIDIA/AMD only | `comfyui: container-gpu` |
| `disabled` | **Default** | — |

## 2. Two findings that invert the obvious approach

1. **Unsloth trains on Apple Silicon now.** The "CUDA-only, Triton kernels" understanding is stale. Upstream: *"macOS: Training, MLX and GGUF inference are ALL supported."*
2. **Its Docker image does not.** `unsloth/unsloth:latest` publishes **`linux/amd64` only** — no arm64 manifest. On Apple Silicon the container is emulated and GPU-less, i.e. useless for training.

So on this stack's primary hardware the *host* flavours are the real product and the container is speculative. Same conclusion ComfyUI reached, for the same reason.

## 3. ⚠️ Security — this outranks every other consideration

Upstream, verbatim:

> Server-side tools (web search, Python and terminal code execution) run as your user and are on by default. Anyone who can reach the server with the API key can run code on this machine.

Atlas's normal reflex is to route a web surface through Kong and publish it on `HOST_BIND_IP`. **Doing that here publishes arbitrary code execution as the host user.** Non-negotiables for any integration:

1. **Loopback only.** Never `-H 0.0.0.0`; **no default Kong route**. For the `localhost` flavour, `doctor` should *warn* if it detects a wildcard bind rather than silently trust it.
2. **`--disable-tools` by default**, enabling it an explicit documented opt-in.
3. **`UNSLOTH_STUDIO_DISABLE_PUBLIC_CHECK=1`** — on a wildcard bind Studio asks `ifconfig.me` and `check-host.net` about reachability. A self-hosted stack should not phone third parties.
4. **Never `--secure` / `--cloudflare`** — Atlas already ships `cloudflared`; a second vendor-owned tunnel is a surprise egress path.
5. **Admin password from `key_generator.py`**, injected via `UNSLOTH_STUDIO_PASSWORD` (upstream explicitly prefers the env var over `--password VALUE`, which leaks into `ps` and shell history), rotated via `unsloth studio reset-password`.

## 4. Why it is worth doing

Atlas has **no training capability at all** — verified by scanning every `services/*/service.yml`. Ollama, vLLM-Metal, LiteLLM and ComfyUI are inference or serving. The `ml-eng` track carries everything *around* training — MLflow, MinIO, JupyterHub, Zeppelin, Ray, Spark, Label Studio — with nothing in the middle that fine-tunes a model.

```
Label Studio ──► dataset ──► MinIO
                               │
                    Unsloth Studio (Data Recipes → train) ──► MLflow
                               │
                        GGUF export ──► Ollama / vLLM-Metal
                               │
                          LiteLLM ──► Langfuse
```

The exported GGUF is consumed by a serving path Atlas already owns, so this closes a loop rather than bolting on.

## 5. Unsloth Studio is the training surface

Not a chat skin: **Data Recipes** (datasets from PDFs/CSVs/DOCX), LoRA/QLoRA/full fine-tuning/pretraining, GRPO/DPO/FP8 RL, GGUF/NVFP4/FP8 export, and an OpenAI-compatible server (`/v1/chat/completions`, `/v1/models`, plus Anthropic `/v1/messages`). Launch is `unsloth studio -p <port>`; it **binds `127.0.0.1` by default**, which matches Atlas's managed-host doctrine without coaxing.

Three upstream affordances make it unusually well-suited to Atlas management:

| Affordance | Why it matters |
|---|---|
| `UNSLOTH_STUDIO_HOME` | Isolated venv, `auth/`, `studio.db`, cache and llama.cpp build in one dir → maps onto `UNSLOTH_STATE_DIR`, makes `remove` a directory delete |
| `./install.sh --local` from a git clone | Atlas can **pin a SHA** instead of piping the vendor's `curl \| sh` — same discipline as ComfyUI custom nodes |
| `UNSLOTH_STUDIO_PASSWORD` | Documented non-interactive path for headless setup |

## 6. Proposed service shape

```
UNSLOTH_SOURCE, UNSLOTH_LOCALHOST_PORT, UNSLOTH_ENDPOINT, UNSLOTH_SCALE,
UNSLOTH_STATE_DIR, UNSLOTH_PIN_REF, UNSLOTH_MODELS_PATH,
UNSLOTH_MIN_MEMORY_GB, UNSLOTH_MIN_DISK_GB,
UNSLOTH_ADMIN_PASSWORD (secret), UNSLOTH_API_KEY (secret),
UNSLOTH_ENABLE_TOOLS (default false)
```

Lifecycle matching existing managed hosts:
`./start.sh unsloth preflight|install|start|stop|status|health|remove`

`category: llm`, `ml-eng` track. Like ComfyUI-MPS and vLLM-Metal it is a **host-global singleton shared across consumers**, so `./stop.sh` must leave it running and report it — `report_managed_hosts_left_running` already exists for this.

## 7. Integration surface

| Peer | Wiring | Confidence |
|---|---|---|
| LiteLLM | Register the OpenAI endpoint as an upstream, like vLLM-Metal. Bearer `sk-unsloth-…` | High |
| MinIO | Datasets in, adapters/GGUF out | High |
| Ollama / vLLM-Metal | Serve the exported GGUF | High |
| Langfuse | Free once served through LiteLLM (gateway-level) | High |
| Hermes / OpenClaw | Upstream ships `unsloth start hermes` and `unsloth start openclaw` — **both are Atlas services** | High |
| MLflow | Built on HF TRL, so autologging is the likely hook | Medium — unverified |
| Kong | **Deliberately none by default** (§3) | — |

## 8. Risks

- **GPU contention — gating.** Fine-tuning is the heaviest GPU workload there is, and this stack has documented contention: concurrent stacks sharing host-global GPU runtimes have caused machine-level panics. Training beside a resident 18 GB `qwen3.8:latest`, ComfyUI holding Krea 2 and vLLM-Metal is a unified-memory exhaustion event waiting to happen. **Treat aggregate-residency admission control as a prerequisite, not a companion.**
- **Long-running jobs vs Atlas's lifecycle.** Atlas services are start/stop daemons; a fine-tune is an hours-long *job* with artifacts. `./stop.sh` mid-run destroys work.
- **Licensing.** Core is Apache-2.0 (matching Atlas); **Studio UI is AGPL-3.0** — and §5 makes Studio the centrepiece, so this is load-bearing rather than incidental. Atlas does not redistribute upstream artifacts, but a consumer shipping Atlas as a product should know one component differs.
- **Jupyter overlap (container flavour only).** The image bundles Jupyter on `8888` and expects `JUPYTER_PASSWORD`; Atlas already runs JupyterHub.
- **Nightly-by-default install.** `./install.sh --local` builds from `main`. Atlas must pin `UNSLOTH_PIN_REF` and treat updates as deliberate.

## 9. Phased plan

| Phase | Deliverable | Gate |
|---|---|---|
| **0** | GPU aggregate-residency admission control | Do not add the heaviest GPU consumer to an unbounded stack |
| **1** | `localhost` flavour: manifest, endpoint wiring, LiteLLM registration, `doctor` checks for tools-enabled / wildcard-bind | Cheapest slice, zero install risk, proves the contract |
| **2** | `managed-localhost`: manager module, pinned install, state dir, generated password, lifecycle verbs | `status`/`health` clean without a training run |
| **3** | Train → export GGUF → serve via Ollama → route via LiteLLM, end to end | A fine-tuned adapter answers through the gateway |
| **4** | `container-gpu` for Linux/NVIDIA hosts | Only if a consumer has that hardware |
| **5** | MLflow autologging; Ray/Airflow job scheduling | Only after Phase 3 proves the loop |

`localhost` is deliberately Phase 1: it needs no installer, so it validates the endpoint contract, the security posture and the LiteLLM wiring before Atlas takes on lifecycle ownership.

## 10. Decisions needed

1. **Is Phase 0 a prerequisite?** The central judgment call.
2. **Job model** — Studio-owned, or Ray/Airflow-scheduled?
3. **Tools on or off by default?** Off is safe but disables Studio's agentic features.
4. **Container flavour at all**, on Apple-Silicon-only infrastructure?
5. **Track membership** — `ml-eng` only, or `gen-ai-eng` too?

## 11. Evidence log

| Claim | How verified |
|---|---|
| macOS training + MLX + Metal | Upstream README, hardware-support section |
| Image is amd64-only | Docker registry manifest query — `linux/amd64` only, no arm64 |
| Studio = training surface | README feature list + Studio section |
| Binds `127.0.0.1` by default | README remote-access section |
| Tools run as your user, on by default | README security paragraph (quoted in §3) |
| `UNSLOTH_STUDIO_HOME` isolation | README advanced-install section |
| `UNSLOTH_STUDIO_PASSWORD` automation | README headless-setup guidance |
| Reachability check contacts third parties | README wildcard-bind paragraph |
| Apache-2.0 core / AGPL-3.0 Studio | GitHub API `spdx_id` + README licence section |
| Atlas has no training service | Manifest scan across `services/*/service.yml` |
| Dual-flavour precedent | `services/comfyui/service.yml` ships `localhost` *and* `managed-localhost-mps` |

**Not verified, flagged as such:** MLflow autologging behaviour; real memory footprint of a training run on this hardware; whether the container's bundled Jupyter can be cleanly disabled; and whether Studio exposes a non-interactive way to mint the `sk-unsloth-…` API key that LiteLLM registration needs.

A fuller spec version of this study is drafted at `docs/superpowers/specs/2026-08-11-unsloth-service-feasibility.md` but is **not yet committed** — say the word and it can land via gitflow.
